devito/dimension.py

```diff
@@ -287,33 +287,50 @@ class SubDimension(DerivedDimension):
 
     :param name: Name of the dimension symbol.
     :param parent: Parent dimension from which the SubDimension is created.
-    :param lower: Lower offset from the ``parent`` dimension.
-    :param upper: Upper offset from the ``parent`` dimension.
+    :param lower: Symbolic expression to provide the lower bound
+    :param upper: Symbolic expression to provide the upper bound
     """
 
     def __new__(cls, name, parent, lower, upper, **kwargs):
         newobj = DerivedDimension.__new__(cls, name, parent, **kwargs)
-        newobj._lower = lower
-        newobj._upper = upper
+        newobj._interval = sympy.Interval(lower, upper)
         return newobj
 
+    @classmethod
+    def left(cls, name, parent, thickness):
+        return cls(name, parent,
+                   lower=parent.symbolic_start,
+                   upper=parent.symbolic_start+thickness-1)
```

Comment:
this is taking -1, but ``middle`` is not. Is this consistent/expected?

Reply:
The bounds are inclusive now, right?  So (with square brackets denoting closed set, not python notation):
left = [0,thickness-1]  (e.g. 0,1,2,3)
middle = [thickness, end-thickness] (e.g 4,5,6)
right = [end-thickness+1, end] (e.g. 7,8,9,10)

Reply:
fine, right